### ISSUE TITLE
Loading keyboard layout from xkb files

### ISSUE THREAD

**Issue by Isti115** [riverwm/river#869]
I have noticed that the support for loading custom keyboard layout from files is currently missing, so I looked into the source and found that you have marked it as a `TODO`.

I managed to get it working with the following patch:
```patch
Submodule deps/zig-xkbcommon contains modified content
diff --git a/deps/zig-xkbcommon/src/xkbcommon.zig b/deps/zig-xkbcommon/src/xkbcommon.zig
index cb81158..ae1ad3b 100644
--- a/deps/zig-xkbcommon/src/xkbcommon.zig
+++ b/deps/zig-xkbcommon/src/xkbcommon.zig
@@ -112,9 +112,11 @@ pub const Keymap = opaque {
     extern fn xkb_keymap_new_from_names(context: *Context, names: ?*const RuleNames, flags: CompileFlags) ?*Keymap;
     pub const newFromNames = xkb_keymap_new_from_names;
 
-    // TODO
-    //extern fn xkb_keymap_new_from_file(context: *Context, file: *FILE, format: Format, flags: CompileFlags) ?*Keymap;
-    //pub const newFromFile = xkb_keymap_new_from_file;
+    pub const FILE = opaque {};
+
+    extern fn xkb_keymap_new_from_file(context: *Context, file: *FILE, format: Format, flags: CompileFlags) ?*Keymap;
+    pub const newFromFile = xkb_keymap_new_from_file;
 
     extern fn xkb_keymap_new_from_string(context: *Context, string: [*:0]const u8, format: Format, flags: CompileFlags) ?*Keymap;
     pub const newFromString = xkb_keymap_new_from_string;
diff --git a/river/command/keyboard.zig b/river/command/keyboard.zig
index 93f5cb9..c089ea7 100644
--- a/river/command/keyboard.zig
+++ b/river/command/keyboard.zig
@@ -26,12 +26,15 @@ const util = @import("../util.zig");
 const Error = @import("../command.zig").Error;
 const Seat = @import("../Seat.zig");
 
+extern fn fopen(__filename: [*c]const u8, __modes: [*c]const u8) *xkb.Keymap.FILE;
+
 pub fn keyboardLayout(
     _: *Seat,
     args: []const [:0]const u8,
     _: *?[]const u8,
 ) Error!void {
     const result = flags.parser([:0]const u8, &.{
+        .{ .name = "file", .kind = .arg },
         .{ .name = "rules", .kind = .arg },
         .{ .name = "model", .kind = .arg },
         .{ .name = "variant", .kind = .arg },
@@ -39,23 +42,35 @@ pub fn keyboardLayout(
     }).parse(args[1..]) catch {
         return error.InvalidValue;
     };
-    if (result.args.len < 1) return Error.NotEnoughArguments;
-    if (result.args.len > 1) return Error.TooManyArguments;
 
-    const rule_names = xkb.RuleNames{
-        .layout = result.args[0],
-        // TODO(zig) these should eventually coerce without this hack.
-        .rules = if (result.flags.rules) |s| s else null,
-        .model = if (result.flags.model) |s| s else null,
-        .variant = if (result.flags.variant) |s| s else null,
-        .options = if (result.flags.options) |s| s else null,
+    const new_keymap = if (result.flags.file) |path| file: {
+        var file = fopen(path.ptr, "r");
+
+        break :file xkb.Keymap.newFromFile(
+            server.config.xkb_context,
+            file,
+            .text_v1,
+            .no_flags,
+        ) orelse return error.InvalidValue;
+    } else rule: {
+        if (result.args.len < 1) return Error.NotEnoughArguments;
+        if (result.args.len > 1) return Error.TooManyArguments;
+
+        const rule_names = xkb.RuleNames{
+            .layout = result.args[0],
+            // TODO(zig) these should eventually coerce without this hack.
+            .rules = if (result.flags.rules) |s| s else null,
+            .model = if (result.flags.model) |s| s else null,
+            .variant = if (result.flags.variant) |s| s else null,
+            .options = if (result.flags.options) |s| s else null,
+        };
+        break :rule xkb.Keymap.newFromNames(
+            server.config.xkb_context,
+            &rule_names,
+            .no_flags,
+        ) orelse return error.InvalidValue;
     };
 
-    const new_keymap = xkb.Keymap.newFromNames(
-        server.config.xkb_context,
-        &rule_names,
-        .no_flags,
-    ) orelse return error.InvalidValue;
     defer new_keymap.unref();
 
     server.config.keymap.unref();
```

With this applied I can run `riverctl keyboard-layout -file ~/.config/xkb/Vortex-Core.xkb` and have my custom layout working without disturbing other functionality, but I'm not entirely happy with it, as utilizing `fopen` and the opaque `FILE` definition seems like a workaround instead of a solution.

Initially I have tried opening the file like this:
```zig
var file = std.fs.cwd().openFile(path, .{}) catch {
    return error.Other;
};
defer file.close();
```
And calling `xkb.Keymap.newFromFile` with `&file.handle`, while parametrizing
`xkb_keymap_new_from_file` as `file: *std.fs.File.Handle`, which compiled fine, but when I tried to apply the layout through `riverctl` I got an _"Invalid value"_ error and found the following runtime error in the log:
> Couldn't read XKB file (unknown file): Bad file descriptor

Do you have any idea what's wrong with this approach? If so, I'd be happy to fix it and submit it as a PR.

---

_ps.: Yesterday was my first day writing Zig, so I appreciate any constructive criticism regarding the code!_

**Comment by Leon-Plickat** [comment]
I am not sure I follow what your issue is. I use a custom xkb layout and it works fine without any extra support from river. Note that you need to put your layout into `$HOME/.xkb/symbols/` for xkb to load it.

**Comment by Isti115** [comment]
@Leon-Plickat Hmm, after reading your comment I have tried all the ways I could think of, but couldn't load my custom layout through the existing `keyboard-layout` command, not matter where I placed it or how I named it. :thinking: 

(I also took a look at your article, and had severe flashbacks to the spring of 2020 when I was figuring this out for myself... 😄 The frustration of not having any proper documentation and the hours of trial and error were real! Here's what I'm using currently: https://github.com/Isti115/dotfiles/blob/master/.config/xkb/Vortex-Core.xkb [The included files are in the neighboring directories.])

Anyway, I have submitted a PR, so I'll close this issue.
